### PR TITLE
pr-auditor: test plan checks and audit trail for exceptions

### PR DESCRIPTION
**Pull request by bobheadxi**
Introduces a bot to enforce the following:

1. Author creates a PR. Template asks them to provide a [test plan](https://docs.sourcegraph.com/dev/background-information/testing_principles#test-plans)
2. On PR edit, auditor GitHub Action runs, setting a status if no test plan is detected (see example below)
2. On PR merge, auditor GitHub Action runs:
	a. If test plan is provided and reviewed, do nothing
	b. If test plan and/or review is not provided, create issue in `sourcegraph/sec-audit-trail` with request for explanation from PR merger. The author should close the issue when an explanation is provided. A status is also created on the commit linking to the issue.

Closes https://github.com/sourcegraph/sourcegraph/issues/29765 , part of https://github.com/sourcegraph/sourcegraph/issues/29764.

Relevant guidance: https://github.com/sourcegraph/sourcegraph/pull/30789 , now https://docs.sourcegraph.com/dev/background-information/testing_principles#test-plans

### Test plan

Nothing rigorous needed, but I've included unit tests to validate the test plan detection. In practice, not 100% sure this will work with the built-in actions token. Will probably need to follow up after merge with fixes after seeing this live, but right now it seems to be ~working

image

(details links to test plan docs!)

### CODE REVIEW

**Review by jhchabran — COMMENTED** [review]
Looks great 👌! Let's keep going :) 

Could we parameterize the labels of the issues and tag them automatically? It will make sure that the issues are not becoming a mess in the future if we log other things.

**Comment by jhchabran** [comment]
@bobheadxi I went ahead and created https://github.com/sourcegraph/sec-pr-audit-trail (with only access to us for now).

**Review by bobheadxi — COMMENTED** [review]
@jhchabran could we just make https://github.com/sourcegraph/sec-pr-audit-trail public as discussed in the RFC? Half the explanations will be in the already-public PRs by default anyway

> Could we parameterize the labels of the issues and tag them automatically? It will make sure that the issues are not becoming a mess in the future if we log other things.

What labels would you like to see?

**Comment by jhchabran** [comment]
> could we just make sourcegraph/sec-pr-audit-trail public as discussed in the RFC?

I have mixed feelings on this one. On one hand, as you mentioned, stuff is public anyway. On the other hand, it would make the audit trail unusable for other changes in any private repository. And the value for bystanders is rather low. 

A possible good way to start is in the open and revert if find ourselves in situations where that is problematic. Thanks for defending the tough position of keeping things public by default 💪!

> What labels would you like to see?

- `sourcegraph` (so we can filter by repo) 
- `acceptance` (so we can still filter things in the future if we log additional items in there)

**Comment by bobheadxi** [comment]
> A possible good way to start is in the open and revert if find ourselves in situations where that is problematic. Thanks for defending the tough position of keeping things public by default 💪!

yes, let's start open! 😁 

> And the value for bystanders is rather low.

The direct value for bystanders of a company being open source is often pretty low: in most open-source companies the vast majority of contributors are employees, and even now we rarely receive contributions externally. A lot of our handbook content is entirely unhelpful too (who needs kubectl access to k8s.sgdev.org?) - so it's more about the principle that we are open by default, and the trust we hope to gain by being transparent: https://handbook.sourcegraph.com/company-info-and-process/values#open-and-transparent So the question is less "what is the value for bystanders?" and "why should we _not_ be open in this case?" And at the moment I don't see much of a reason, given our core repo is open, and I think it's a cool gesture to make that our SOC2 compliance is publicly auditable by anyone, similar to how an open source project's source code can be audited by anyone

> On the other hand, it would make the audit trail unusable for other changes in any private repository

Links to PRs in private repositories are frequently posted already, so that's not too much of a concern. It's possible explanations could leak details, but we can either a) add warnings to not add sensitive content to explanations or b) open another audit trail repo for private repos. I think a) should be sufficient, but we can revisit this when needed :)

**Comment by bobheadxi** [comment]
I'll get the labels and https://github.com/sourcegraph/sourcegraph/issues/30396#issuecomment-1026028932 impleemnted

**Comment by jhchabran** [comment]
The sourcegraph/sec-pr-audit-trail repo is now public :)

File: dev/pr-auditor/issue.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+		issueTitle      string
+		issueBody       string
+		issueAssignees  = []string{}
+	)
+
+	if !result.Reviewed {
+		exceptionLabels = append(exceptionLabels, "exception/review")
+	}
+	if !result.HasTestPlan() {
+		exceptionLabels = append(exceptionLabels, "exception/test-plan")
```
Comment (jhchabran) [inline dev/pr-auditor/issue.go:21]: That's a great idea 👍

Comment (jhchabran) [inline dev/pr-auditor/issue.go:21]: That's a great idea 👍

**Review by jhchabran — APPROVED** [review]
LGTM!

**Comment by bobheadxi** [comment]
Merging this for now so I can keep an eye on how it behaves this afternoon!

**Pull request merged**